### PR TITLE
Make trainer.state a read-only property

### PR DESCRIPTION
**Pull request by zerogerc**
## What does this PR do?

Follow-up on https://github.com/PyTorchLightning/pytorch-lightning/pull/2541 to make trainer state read-only.

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
good work, but I would rather add also setter which main checks that the values in TrainerState type :]

**Comment by justusschock** [comment]
@Borda I would prefer to keep it read-only as well. everything that is exposed will most likely be used by users in some way. And checking for valid arguments is quite hard. We can check for types, but even arguments with correct types may be wrong  w.r.t. to trainer's flow.

File: pytorch_lightning/trainer/states.py [hunk 1]
```diff
@@ -42,20 +42,20 @@ def wrapped_fn(self, *args, **kwargs):
...
-            if self.state == TrainerState.INTERRUPTED:
+            if self._state == TrainerState.INTERRUPTED:
                 return result
 
             if exiting is not None:
-                self.state = exiting
+                self._state = exiting
             else:
-                self.state = state_before
+                self._state = state_before
```
Comment (Borda) [inline pytorch_lightning/trainer/states.py:58]: `self._state = exiting if exiting is not None else state_before `

**Review by justusschock — APPROVED** [review]
To me this looks completely fine :D 

Making things, the user isn't expected to change directly read-only is always a good idea I think :D

**Review by SkafteNicki — APPROVED** [review]
LGTM

**Review by Borda — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
I like to see this ❤️

File: pytorch_lightning/trainer/states.py (hunk 1 above)
Comment (zerogerc) [reply pytorch_lightning/trainer/states.py]: Thanks for fixing this =)

**Pull request merged**